xds/internal/balancer/outlierdetection/balancer.go

```diff
@@ -105,3 +135,736 @@ func (bb) ParseConfig(s json.RawMessage) (serviceconfig.LoadBalancingConfig, err
+		return
+	}
+	if ei, ok := b.child.(balancer.ExitIdler); ok {
+		b.childMu.Lock()
+		defer b.childMu.Unlock()
+		ei.ExitIdle()
+		return
+	}
+
+	// Fallback for children handled in clusterimpl balancer Removing SubConns
+	// is defined in API and also in graceful switch balancer, but already done
+	// in ClusterImpl. I guess we should do that here?
+}
+
+// wrappedPicker delegates to the child policy's picker, and when the request
+// finishes, it increments the corresponding counter in the map entry referenced
+// by the subConnWrapper that was picked. If both the `success_rate_ejection`
+// and `failure_percentage_ejection` fields are unset in the configuration, this
+// picker will not count.
+type wrappedPicker struct {
+	childPicker balancer.Picker
+	noopPicker  bool
+}
+
+func (wp *wrappedPicker) Pick(info balancer.PickInfo) (balancer.PickResult, error) {
+	pr, err := wp.childPicker.Pick(info)
+	if err != nil {
+		return balancer.PickResult{}, err
+	}
+
+	done := func(di balancer.DoneInfo) {
+		if !wp.noopPicker {
+			incrementCounter(pr.SubConn, di)
+		}
+		if pr.Done != nil {
+			pr.Done(di)
+		}
+	}
+	scw, ok := pr.SubConn.(*subConnWrapper)
+	if !ok {
+		// This can never happen, but check is present for defensive
+		// programming.
+		return balancer.PickResult{
+			SubConn: pr.SubConn,
+			Done:    done,
+		}, nil
+	}
+	return balancer.PickResult{
+		SubConn: scw.SubConn,
+		Done:    done,
+	}, nil
+}
+
+func incrementCounter(sc balancer.SubConn, info balancer.DoneInfo) {
+	scw, ok := sc.(*subConnWrapper)
+	if !ok {
+		// Shouldn't happen, as comes from child
+		return
+	}
+
+	// scw.addressInfo and callCounter.activeBucket can be written to
+	// concurrently (the pointers themselves). Thus, protect the reads here with
+	// atomics to prevent data corruption. There exists a race in which you read
+	// the addressInfo or active bucket pointer and then that pointer points to
+	// deprecated memory. If this goroutine yields the processor, in between
+	// reading the addressInfo pointer and writing to the active bucket,
+	// UpdateAddresses can switch the addressInfo the scw points to. Writing to
+	// an outdated addresses is a very small race and tolerable. After reading
+	// callCounter.activeBucket in this picker a swap call can concurrently
+	// change what activeBucket points to. A50 says to swap the pointer, but I
+	// decided to make create new memory for both active and inactive bucket,
+	// and have this race instead write to deprecated memory. If you swap the
+	// pointers, this write would write to the inactive buckets memory, which is
+	// read throughout in the interval timers algorithm.
+	addrInfo := (*addressInfo)(atomic.LoadPointer(&scw.addressInfo))
+	if addrInfo == nil {
+		return
+	}
+	ab := (*bucket)(atomic.LoadPointer(&addrInfo.callCounter.activeBucket))
+
+	if info.Err == nil {
+		atomic.AddUint32(&ab.numSuccesses, 1)
+	} else {
+		atomic.AddUint32(&ab.numFailures, 1)
+	}
+	atomic.AddUint32(&ab.requestVolume, 1)
```

Comment:
Remove this field. Just add numSuccesses and numFailures and then there is one fewer race.